wrapper/src/main/java/software/amazon/jdbc/util/storage/StorageServiceImpl.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.util.storage;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Logger;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import software.amazon.jdbc.util.ItemDisposalFunc;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.ShouldDisposeFunc;
+
+public class StorageServiceImpl implements StorageService {
+  private static final Logger LOGGER = Logger.getLogger(StorageServiceImpl.class.getName());
+  protected static Map<String, ExpirationCache<Object, ?>> caches = new ConcurrentHashMap<>();
+
+  public StorageServiceImpl() {
+
+  }
+
+  @Override
+  public <V> void registerItemCategoryIfAbsent(
+      String itemCategory,
+      Class<V> itemClass,
+      boolean isRenewableExpiration,
+      long timeToLiveNanos,
+      long cleanupIntervalNanos,
+      @Nullable ShouldDisposeFunc<V> shouldDisposeFunc,
+      @Nullable ItemDisposalFunc<V> itemDisposalFunc) {
+    caches.computeIfAbsent(
+        itemCategory,
+        category -> new ExpirationCache<>(
```

Comment:
So the design is to have a separate cache (that sits in another cache) for each item category?

Reply:
Each item category has its own cache so that each item category can have its own settings, since different item categories will have different needs (expiration time, cleanup time, shouldDisposeFunc, itemDisposalFunc). The caches are stored in a map that serves as a cache directory so we can quickly find the given cache for an item category.

Reply:
Map of caches might not be a savvy solution. Each cache has its own clean-up thread so I can assume that it would be as many clean-up threads as data types. Most probably each cache has just one or two entries. For example, topology cache will have a single entry. If user app works with two different clusters (that is a rare case) then topology cache will have two entries... Can we pack all entries into a single cache (i.e. instead of map of caches use a single cache)?

Reply:
Good point, but it sounds like the main issue is the cleanup mechanism and number of threads instead of the internal structure. I'd like to keep the data grouped together because it allows us to cleanly store the data together with the settings for how to handle that data in one spot (ExpiringCache). I've updated the code now so that there is only one cleanup thread for all items, which is created and run in StorageServiceImpl.